sdks/java/core/src/main/java/org/apache/beam/sdk/testing/BigqueryMatcher.java

```diff
@@ -0,0 +1,159 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * A matcher to verify data in BigQuery by processing given query
+ * and comparing with content's checksum.
+ *
+ * <p>Example:
+ * <pre>{@code [
+ *   assertTrue(job, new BigQueryMatcher(appName, projectId, queryString, expectedChecksum));
+ * ]}</pre>
+ */
+public class BigqueryMatcher extends TypeSafeMatcher<PipelineResult>
+    implements SerializableMatcher<PipelineResult> {
+  private static final Logger LOG = LoggerFactory.getLogger(BigqueryMatcher.class);
+
+  private final String applicationName;
+  private final String projectId;
+  private final String query;
+  private final String expectedChecksum;
+  private String actualChecksum;
+
+  public BigqueryMatcher(
+      String applicationName, String projectId, String query, String expectedChecksum) {
+    validateArguments(applicationName, projectId, query, expectedChecksum);
+
+    this.applicationName = applicationName;
+    this.projectId = projectId;
+    this.query = query;
+    this.expectedChecksum = expectedChecksum;
+  }
+
+  @Override
+  protected boolean matchesSafely(PipelineResult pipelineResult) {
+    LOG.info("Verifying Bigquery data");
+    Bigquery bigqueryClient = newBigqueryClient(applicationName);
+
+    QueryResponse response;
+    try {
+      LOG.info("Executing query: {}", query);
+      QueryRequest queryContent = new QueryRequest();
+      queryContent.setQuery(query);
+      response = bigqueryClient.jobs().query(projectId, queryContent).execute();
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to retrieve BigQuery data.", e);
+    }
+
+    if (response == null) {
+      LOG.warn("Invalid query response.");
+      return false;
+    }
+    actualChecksum = hashing(response.getRows());
+    LOG.info("Generated a SHA1 checksum based on queried data: {}", actualChecksum);
+    return expectedChecksum.equals(actualChecksum);
+  }
+
+  void validateArguments(String... args) {
+    for (String argument : args) {
+      checkArgument(
+          !Strings.isNullOrEmpty(argument),
+          "Expected valid argument, but was %s", argument);
+    }
+  }
+
+  Bigquery newBigqueryClient(String applicationName) {
+    HttpTransport transport = Transport.getTransport();
+    JsonFactory jsonFactory = Transport.getJsonFactory();
+    Credential credential = getDefaultCredential(transport, jsonFactory);
+
+    return new Bigquery.Builder(transport, jsonFactory, credential)
+        .setApplicationName(applicationName)
+        .build();
+  }
+
+  Credential getDefaultCredential(HttpTransport transport, JsonFactory jsonFactory) {
+    GoogleCredential credential;
+    try {
+      credential = GoogleCredential.getApplicationDefault(transport, jsonFactory);
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to get application default credential.", e);
+    }
+
+    if (credential.createScopedRequired()) {
+      Collection<String> bigqueryScope = BigqueryScopes.all();
```

Comment:
right, BigqueryScopes.CLOUD_PLATFORM_READ_ONLY seems enough.

According to: https://developers.google.com/resources/api-libraries/documentation/bigquery/v2/java/latest/com/google/api/services/bigquery/BigqueryScopes.html#CLOUD_PLATFORM_READ_ONLY